server/location-model.js

```diff
@@ -0,0 +1,42 @@
+const db = require("./db");
+const log = require("./logger")("location-model");
+const { hash } = require("./util");
+
+const Location = db.define("location", {
+  location: {
+    type: db.Sequelize.STRING,
+    unique: true,
+    allowNull: false,
+    set(val) {
+      this.setDataValue("location", hash(val));
+    }
+  },
+  votes: {
+    type: db.Sequelize.INTEGER,
+    defaultValue: 0
+  }
+});
+
+Location.findOneWithLocation = function findOneWithLocation(location) {
+  return Location.findOne({ where: { location: hash(location) } });
+};
+
+Location.upvoteOrCreate = function upvoteOrCreate(location) {
+  log.debug(`Incrementing location ${location}`);
+  const where = { location: hash(location) };
+  return Location.increment("votes", { where }).then(resp => {
+    // not entirely sure what sequelize returns, but this works:
```

Comment:
You're applying the increment on possibly multiple locations.
So you get back an array of results, each result has the standard sequelize response `[rows, result]`, where `rows` will always be `undefined` because increment does not return the affected rows, and `success` is an integer 1 or 0.

```js
  return Location.findOne({
      where
    })
    .then(locationItem => {
      if (locationItem) {
        log.debug(`Increment returned successful response for ${location}`);
        return locationItem.increment('votes');
      }
      log.debug(
        `Increment returned unsuccessful response. Creating new location ${location}.`
      );
      return Location.create({
        location,
        votes: 1
      });
    })
```

Reply:
Noice! Thanks for the explanation. I kind of like only making the one database query. But perhaps it would be good to ensure we're actually incrementing the correct location by fetching it first. 🤔